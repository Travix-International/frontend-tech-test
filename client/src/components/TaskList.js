import React, { Component } from 'react';
import { connect } from 'react-redux';
import { fetchAll, updateTask, deleteTask } from '../actions/tasks';
import TaskItem from './TaskItem';

import '../style/TaskList.scss';

class TaskList extends Component {
  constructor() {
    super();

    this.handleUpdate = this.handleUpdate.bind(this);
    this.handleRemove = this.handleRemove.bind(this);
  }
  componentDidMount() {
    this.props.fetchAll();
  }

  handleUpdate(task) {
    const { updateTask } = this.props;
    updateTask(task);
  }

  handleRemove(task) {
    const { deleteTask } = this.props;

    deleteTask(task.id);
  }

  renderTaskList() {
    const { tasks } = this.props;

    if (tasks.length === 0) {
      return (
        <div className="msg">
          You not have any task,<br />
          click on add task button to add one
        </div>
      );
    }

    return tasks.map((task, index) => {
      return (<TaskItem
        key={index}
        task={task}
        handleUpdate={this.handleUpdate}
        handleRemove={this.handleRemove} />);
    });
  }

  render() {
    return (
      <div>
        <ul className="TaskList">
          {this.renderTaskList()}
        </ul>
      </div>
    );
  }
}

const mapStateToProps = state => {
  const { tasks } = state;
  return tasks;
};

const mapDispatchToProps = dispatch => {
  return {
    fetchAll: () => {
      dispatch(fetchAll());
    },
    updateTask: (task) => {
      dispatch(updateTask(task));
    },
    deleteTask: (id) => {
      dispatch(deleteTask(id));
    }
  };
};

const ConnectedTaskList = connect(
  mapStateToProps,
  mapDispatchToProps
)(TaskList);

export default ConnectedTaskList;
