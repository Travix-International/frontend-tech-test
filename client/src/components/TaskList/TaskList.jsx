import React from 'react';
import PropTypes from 'prop-types';

import { TaskListType } from '../../constants/propTypes';
import styles from './TaskList.module.scss';
import Task from '../Task/Task';

const TaskList = ({ tasks, editTask, deleteTask }) => (
  <div className={styles.TaskList}>
    {tasks.map(task => (
      <Task key={task.id} task={task} editTask={editTask} deleteTask={deleteTask} />
    ))}
  </div>
);

TaskList.propTypes = {
  tasks: TaskListType,
  editTask: PropTypes.func.isRequired,
  deleteTask: PropTypes.func.isRequired,
};

TaskList.defaultProps = {
  tasks: []
}

export default TaskList;
