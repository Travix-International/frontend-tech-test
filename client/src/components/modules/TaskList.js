import React, {Component} from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import Loader from 'components/modules/Loader'
import TaskActive from 'components/modules/TaskActive'
import TaskCompleted from 'components/modules/TaskCompleted'
import TaskCreate from 'components/modules/TaskCreate'
import { toggleTask, fetchTasks, updateTask, createTask, deleteTask } from 'actions/Tasks'

export class TaskList extends Component {
  constructor(props) {
    super(props);

    this.state = {
      newTitle: '',
    }

    this.onTaskDelete = this.onTaskDelete.bind(this);
    this.onTaskToggle = this.onTaskToggle.bind(this);
    this.onTaskUpdate = this.onTaskUpdate.bind(this);
    this.onTaskCreate = this.onTaskCreate.bind(this);
    this.renderActiveTasks = this.renderActiveTasks.bind(this);
    this.renderCompleteTasks = this.renderCompleteTasks.bind(this);
  }

  componentDidMount() {
    setTimeout(() => this.props.fetchTasks(), 1000);
    setInterval(() => this.props.fetchTasks(), 30000);
  }

  onTaskToggle({ id }) {
    const updatedTasks = this.props.tasks.tasks.map(item => {
      item.status = item.id === id ? !item.status : item.status;
      return item;
    });
    return this.props.toggleTask(updatedTasks, id)
  }

  onTaskDelete({ id }) {
    const updatedTasks = this.props.tasks.tasks.filter(item => item.id !== id);
    return this.props.deleteTask(updatedTasks, id)
  }

  onTaskUpdate(event, task, target) {

    if (event.type === 'submit') {
      event.preventDefault()
      const { tasks: { tasks, isFetching } } = this.props;
      return this.props.updateTask(tasks, task);
    }

    const updatedTasks = this.props.tasks.tasks.map(item => {
      item[target] = item.id === task.id ? event.target.value : item[target];
      return item;
    });

    if (event.type === 'change') return this.props.updateTask(updatedTasks);

    event.preventDefault()
    return this.props.updateTask(updatedTasks, task);
  }

  onTaskCreate(event, title) {
    if (event.type === 'change' || !title.length) return this.setState({ newTitle: event.target.value });

    event.preventDefault()
    this.setState({ newTitle: '' });
    return this.props.createTask(this.props.tasks.tasks, title);
  }

  renderTasks() {
    const { tasks: { tasks, isFetching } } = this.props;

    if (isFetching) return <Loader />;
    if (!tasks.length) return <div className="no-task">No task created yet...</div>
    return [this.renderActiveTasks(), this.renderCompleteTasks()]
  }

  renderActiveTasks() {
    const { tasks: { tasks } } = this.props;

    const activeTasks = tasks.filter(task => !task.status).sort((a, b) => b.id - a.id);
    return activeTasks.map(task =>
      <TaskActive
        task={task}
        onTaskToggle={this.onTaskToggle}
        onTaskDelete={this.onTaskDelete}
        onTaskUpdate={this.onTaskUpdate}
      />
    )
  }

  renderCompleteTasks() {
    const { tasks: { tasks } } = this.props;
    const completeTasks = tasks.filter(task => task.status).sort((a, b) => b.id - a.id);
    return completeTasks.map(task =>
      <TaskCompleted
        task={task}
        onTaskToggle={this.onTaskToggle}
        onTaskDelete={this.onTaskDelete}
      />
    )
  }

  render() {
    const { newTitle } = this.state

    return (
      <div className={`task-list-container`}>
        <ul className={`task-list`}>
          <TaskCreate title={newTitle} onTaskCreate={this.onTaskCreate}/>
          {this.renderTasks()}
        </ul>
      </div>
    )
  }
}

function mapStateToProps({ tasks }) {
  return { tasks }
}

function mapDespatchToProps(dispatch) {
  return bindActionCreators({ toggleTask, fetchTasks, updateTask, createTask, deleteTask }, dispatch)
}

export default connect(mapStateToProps, mapDespatchToProps)(TaskList);
