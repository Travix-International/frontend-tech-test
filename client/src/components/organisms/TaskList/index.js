import React, { Component } from 'react'
import { connect } from 'react-redux'
import PropTypes from 'prop-types'
import './index.css'

import {
  fetchTasks, updateTask, deleteTask
} from '../../../ducks/modules/tasks'

import TaskItems from '../../molecules/TaskItems/index'

export class TaskList extends Component {
  componentDidMount() {
    this.props.fetchTasks()
  }

  updateTask = (payload) => {
    this.props.updateTask(payload)
  }

  deleteTask = (payload) => {
    this.props.deleteTask(payload)
  }

  render() {
    return (
      <div className='task-list'>
        <TaskItems
          loading={this.props.loading}
          items={this.props.items}
          updateTask={this.updateTask}
          deleteTask={this.deleteTask}
        />
      </div>
    )
  }
}

TaskList.propTypes = {
  items: PropTypes.array.isRequired,
  loading: PropTypes.bool.isRequired,
  fetchTasks: PropTypes.func.isRequired,
  updateTask: PropTypes.func.isRequired,
  deleteTask: PropTypes.func.isRequired
}

const mapStateToProps = (state) => {
  return {
    loading: state.tasks.loading,
    items: state.tasks.items,
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    fetchTasks: () => dispatch(fetchTasks()),
    updateTask: (payload) => dispatch(updateTask(payload)),
    deleteTask: (payload) => dispatch(deleteTask(payload))
  }
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(TaskList)
