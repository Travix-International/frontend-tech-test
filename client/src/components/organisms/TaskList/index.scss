@import 'src/styles/base/medias.scss';

.task-list {
  height: 365px;
  overflow-y: scroll;

  @media #{$tablet} {
    border-radius: 5px;
    width: 50%;
  }

  @media #{$phone} {
    border-radius: 0;
    width: 100%;
  }
}
