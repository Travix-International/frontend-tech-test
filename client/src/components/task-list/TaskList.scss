.ReactTable {
  margin-top: 20px;
  font-weight: 500;
  font-family: "Roboto", "Helvetica", "Arial", sans-serif;
  padding: 10px;
  color:#333; 
}

.ReactTable .rt-tr-group {
  height: 50px;
}

.ReactTable .rt-th, .ReactTable .rt-td {
  padding-top: 15px !important;
}

.checkbox {
  margin-top: -13px !important;
}

.not-found {
  margin-top: 20px !important;
}

th { 
  background: #333; 
  color: #fff !important; 
  font-weight: bold; 
}

td, th {
  padding-left: 0 !important;
}

.button {
  cursor: pointer;
}

.completed {
  text-decoration: line-through;
}

@media only screen and (max-width: 760px),
(min-device-width: 768px) and (max-device-width: 1024px)  {
  table {
    display: block !important;
  }
}