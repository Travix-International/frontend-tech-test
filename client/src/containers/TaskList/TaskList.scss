.list {
  background-color: #fff;
  padding: 1rem;
}
.pagination {
  background-color: #fff;
  text-align: center;
  padding: 1rem;
}
