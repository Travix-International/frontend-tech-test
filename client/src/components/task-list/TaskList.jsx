import React, { useState, memo, Fragment } from 'react';
import { connect } from 'react-redux';
import ReactTable from 'react-table';
import Typography from '@material-ui/core/Typography';
import Checkbox from '@material-ui/core/Checkbox';

import TaskDialog from '../task-dialog/TaskDialog';

import { getTasks, loadTask, deleteTask, loadTasks, editTask } from '../../redux/modules/tasks';

import './TaskList.scss';
import "react-table/react-table.css";

let TaskList = ({
  tasks, loading, loaded, error, dispatch
}) => {
  const [open, setOpen] = useState(false);
  const [id, setId] = useState('');
  if (tasks.length === 0 && (loading || !loaded)) {
    return 'Loading...';
  }

  if (error) {
    return 'An error has occurred';
  }

  if (tasks.length === 0) {
    return (
      <Typography className="not-found" component="h2" variant="h5">
      No tasks found
      </Typography>
    );
  }

  const setComplete = (id, completed) => {
    const task = tasks.find(t => t.id === id);
    const {
      title, description
    } = task;
    
    dispatch(editTask(id, title, description, completed))
      .then(() => dispatch(loadTasks()));
  };

  const handleEdit = (e) => {
    const taskId = e.target.id;

    dispatch(loadTask(taskId))
      .then(() => {
        setId(taskId);
        setOpen(true);
      });
  };

  const handleDelete = (e) => {
    dispatch(deleteTask(e.target.id))
      .then(() => dispatch(loadTasks()));
  };

  return (
    <Fragment>
      {open && (
        <TaskDialog
          id={id}
          isEdit
          open={open}
          setOpen={setOpen}
        />
      )}
      <ReactTable
        className="task-list"
        data={tasks}
        minRows={tasks.length}
        columns={[
          {
            id: "checkbox",
            Header: "",
            accessor: (data) => {
              const completed = data.completed;

              return (
                <Checkbox
                  className="checkbox"
                  checked={completed || false}
                  color={completed ? 'default' : 'primary'}
                  onChange={() => setComplete(data.id, !completed)}
                />
              )
            },
          },
          {
            id: "id",
            Header: () => (
              <div style={{ textAlign:"left" }}>Id</div>
            ),
            accessor: (data) => <div className={data.completed ? 'completed' : ''}>{data.id}</div>,
          },
          {
            id: "title",
            Header: () => (
              <div style={{ textAlign:"left" }}>Title</div>
            ),
            accessor: (data) => <div className={data.completed ? 'completed' : ''}>{data.title}</div>,
          },
          {
            id: "description",
            Header: () => (
              <div style={{ textAlign:"left" }}>Description</div>
            ),
            accessor: (data) => <div className={data.completed ? 'completed' : ''}>{data.description}</div>,
          },
          {
            id: "edit",
            Header: "",
            width: 50,
            accessor: (data) => {
              return (
                <div className="button" onClick={handleEdit}>
                  <img alt="edit" className="icon" src={require('../../images/baseline-edit-24px.svg')} id={data.id} />
                </div>
              )
            },
          },
          {
            id: "delete",
            Header: "",
            width: 50,
            accessor: (data) => {
              return (
                <div className="button" onClick={handleDelete}>
                  <img alt="delete" className="icon" src={require('../../images/baseline-cancel-24px.svg')} id={data.id} />
                </div>
              )
            },
          },
        ]}
        defaultPageSize={10}
        className="-striped -highlight"
      />
    </Fragment>
  );
};

TaskList = memo(TaskList);
TaskList = connect(globalState => ({
  tasks: getTasks(globalState),
  loading: globalState.loading,
  loaded: globalState.loaded,
  error: globalState.error,
}))(TaskList);
export default TaskList;
