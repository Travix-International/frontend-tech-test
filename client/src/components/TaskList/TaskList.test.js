import React from 'react';
import { shallow } from 'enzyme';
import TaskList from './TaskList';

it('renders TaskList correctly', () => {
  const wrapper = shallow(<TaskList deleteTask={() => {}} editTask={() => {}} />);

  expect(wrapper).toMatchSnapshot();
});