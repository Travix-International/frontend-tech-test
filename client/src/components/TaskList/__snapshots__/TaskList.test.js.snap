// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders TaskList correctly 1`] = `ShallowWrapper {}`;
